p2p/src/transport/mconnection.rs

```diff
@@ -0,0 +1,384 @@
+                        public_key,
+                        secret_connection,
+                        stream: stream_clone,
+                    })),
+                    (_, Err(e)) => Some(Err(e)),
+                    (Err(e), _) => Some(Err(e)),
+                }
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+impl Transport for MConnectionTransport {
+    type Connection = MConnection;
+    type Endpoint = MEndpoint;
+    type Incoming = MIncoming;
+
+    /// Creates a new `TcpListener` which will be bound to the specified address.
+    /// The private key will be used to establish a `SecretConnection` each time you connect or
+    /// accept a connection.
+    ///
+    /// See `TcpListener::bind`
+    ///
+    /// # Examples
+    ///
+    /// Creates a TCP listener bound to `127.0.0.1:8080`:
+    ///
+    /// ```no_run
+    /// use rand_core::OsRng;
+    /// use ed25519_dalek::Keypair;
+    ///
+    /// use crate::tendermint_p2p::transport::{BindInfo, Transport};
+    /// use crate::tendermint_p2p::transport::mconnection::MConnectionTransport;
+    ///
+    /// let mut csprng = OsRng{};
+    ///
+    /// let t = MConnectionTransport{};
+    /// let (endpoint, incoming) = t.bind(BindInfo{
+    ///     addr: "127.0.0.1:8080".parse().unwrap(),
+    ///     listen_addrs: vec!["192.168.1.2:26656".parse().unwrap()],
+    ///     private_key: Keypair::generate(&mut csprng),
+    ///
+    /// }).unwrap();
+    /// ```
+    fn bind(&self, bind_info: BindInfo) -> Result<(MEndpoint, MIncoming)> {
+        let listener = TcpListener::bind(bind_info.addr)?;
+        let pk = Arc::new(bind_info.private_key);
+        Ok((
+            MEndpoint {
+                private_key: pk.clone(),
+                // TODO: version should be determined based on the `tendermint` crate version
+                // if we want to support old chains that are using Amino encoding
+                // if not, we should just hardcode the value
+                protocol_version: Version::V0_34,
+                listen_addrs: bind_info.listen_addrs,
+            },
+            MIncoming {
+                tcp_listener: listener,
+                private_key: pk,
+                // TODO: version should be determined based on the `tendermint` crate version
+                // if we want to support old chains that are using Amino encoding
+                // if not, we should just hardcode the value
+                protocol_version: Version::V0_34,
+            },
+        ))
+    }
+
+    /// Noop.
+    fn shutdown(&self) -> Result<()> {
+        // The socket will be closed when the MConnectionTransport is dropped.
+        Ok(())
+    }
+}
+
+impl Read for VirtualStream {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        loop {
+            let mut tmp_buf = Vec::new();
+            // FIXME: we should probably read in loop until decode_length_delimited succeds
```

Comment:
question to all of the team I guess: 

why we're using prost lib and not https://github.com/stepancheg/rust-protobuf, where decode_length_delimited_from_reader exists and we don't need to implement it ourselves?

Reply:
@greg-szabo @thanethomson @tarcieri

Reply:
I can only comment on this from a historical perspective: Prost was originally chosen (back in the KMS) as it was much easier to modify and use to quickly hack together an amino-compatible implementation in Rust for the few messages that the KMS needed to understand. Also, it seemed like the most mature and maintained choice back then.
That said, we kinda sticked to prost even here because it was still nice that you were not forced to use protobuf files and generated (difficult to read) rust code derived from these proto-files. Instead you could just use the derive marco and you wouldn't really need to care about protobuf defineitons. I'm not sure these reasons are still valid (probably not). If there are pros than simply the method `decode_length_delimited_from_reader` then it might be a good time to reconsider choices. if the only reason is the reader method, why not simply implement once and upstream it to prost. But I guess this is very orthogonal to what this PR is trying to achieve and probably deserves its own issue.

Reply:
Yeah it might be worthwhile to revisit this again sometime soon. Does `rust-protobuf` allow for customization of `serde` options at the struct and field-level? Right now we make extensive use of this in our [proto-compiler](https://github.com/informalsystems/tendermint-rs/tree/v0.18.1/tools/proto-compiler) tool.